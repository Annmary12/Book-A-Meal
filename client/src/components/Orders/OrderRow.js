import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import ConfirmCancelOrderModal from './ConfirmCancelOrderModal.js';
import UpdateOrderModal from './UpdateOrderModal.js';
import localizeNum from '../../utils/localizeNum.js';
import formatDateString from '../../utils/formatDateString.js';
import orderIsModifiable from '../../utils/orderIsModifiable.js';
import Cell from '../ResponsiveTable/Cell.js';
import Row from '../ResponsiveTable/Row.js';
import styles from './order.scss';

/**
 * @class OrderRow
 *
 * @param {String} status
 */
class OrderRow extends Component {
  static propTypes = {
    order: PropTypes.shape({
      id: PropTypes.string.isRequired,
      meal: PropTypes.shape({
        id: PropTypes.string.isRequired,
        img: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired
      }).isRequired
    }).isRequired,
    handleCancelOrder: PropTypes.func.isRequired,
    handleUpdateOrder: PropTypes.func.isRequired
  }

  state = {
    cancelModalIsOpen: false,
    updateModalIsOpen: false
  }

  colorStatus = (status) => {
    let color;
    switch (status) {
      case 'pending':
        color = '#f16800';
        break;
      case 'delivered':
        color = 'green';
        break;
      case 'cancelled':
        color = 'red';
        break;
      default:
        color = 'yellow';
    }

    return (<span style={{ color }}>{status}</span>);
  }

  closeCancelConfirmationModal = () => {
    this.setState({ cancelModalIsOpen: false });
  }

  openCancelConfirmationModal = () => {
    this.setState({ cancelModalIsOpen: true });
  }

  closeUpdateModal = () => {
    this.setState({ updateModalIsOpen: false });
  }

  openUpdateModal = () => {
    this.setState({ updateModalIsOpen: true });
  }
  /**
   * @returns {JSX} React JSX
   */
  render() {
    const { order } = this.props;
    return (
      <Row key={order.id}>
        <Cell title="date">
          {formatDateString(order.createdAt)}
        </Cell>
        <Cell title="Meal">
          {order.meal.title}
        </Cell>
        <Cell title="Quantity">
          {order.quantity}
        </Cell>
        <Cell title="Price">
          &#8358;{localizeNum(order.price)}
        </Cell>
        <Cell title="Delivery address">
          <p className={styles.deliveryAddress}>{order.deliveryAddress}</p>
        </Cell>
        <Cell title="Status">
          {this.colorStatus(order.status)}
        </Cell>
        <Cell>
          {
            order.status === 'pending' &&
            orderIsModifiable(order.createdAt) &&
            <Fragment>
              <button
                className={styles.updateBtn}
                onClick={this.openUpdateModal}
              >
                update
              </button>
              <button
                className={styles.deleteBtn}
                onClick={this.openCancelConfirmationModal}
              >
                cancel
              </button>
            </Fragment>
          }
        </Cell>

        <ConfirmCancelOrderModal
          isOpen={this.state.cancelModalIsOpen}
          handleClose={this.closeCancelConfirmationModal}
          orderId={order.id}
          handleCancelOrder={this.props.handleCancelOrder}
        />
        <UpdateOrderModal
          isOpen={this.state.updateModalIsOpen}
          handleClose={this.closeUpdateModal}
          order={order}
          handleUpdateOrder={this.props.handleUpdateOrder}
        />
      </Row>
    );
  }
}

export default OrderRow;
