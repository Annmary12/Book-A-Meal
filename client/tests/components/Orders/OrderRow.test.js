import React from 'react';
import OrderRow from '../../../src/components/Orders/OrderRow';

const props = {
  handleDeliver: jest.fn(),
  order:  {
    "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
    "quantity": 1,
    "price": 500,
    "status": "cancelled",
    "deliveryAddress": "24 irepodun street, bodija, ibadan",
    "phoneNumber": "08131928452",
    "createdAt": "2018-09-10T09:05:53.650Z",
    "updatedAt": "2018-09-10T09:06:06.292Z",
    "meal": {
        "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
        "title": "Beef with rice",
        "description": "Fried rice and beef",
        "price": 500,
        "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
        "caterer": {
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "fullName": "Otse Obande",
            "username": "otseobande",
            "email": "mealbooker@gmail.com",
            "role": "caterer"
        }
    }
  }
}
describe('OrderRow component', () => {
  it('should match snapshot if status is not pending', () => {
    const wrapper = shallow(
      <OrderRow {...props}/>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if status is pending', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'pending'
      }
    };

    const wrapper = shallow(
      <OrderRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if status is delivered', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'delivered'
      }
    };

    const wrapper = shallow(
      <OrderRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot if status is cancelled', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'cancelled'
      }
    };

    const wrapper = shallow(
      <OrderRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });


  it('should match snapshot if status is anything else', () => {
    const updatedProps = {
      ...props,
      order: {
        ...props.order,
        status: 'anything else'
      }
    };

    const wrapper = shallow(
      <OrderRow {...updatedProps} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('closeCancelConfirmationModal method', () => {
    const wrapper = shallow(
      <OrderRow {...props} />
    );
    const action = wrapper.instance();
    it('should set cancelModalIsOpen state to false', () => {
      action.closeCancelConfirmationModal();
      const { cancelModalIsOpen } = action.state;

      expect(cancelModalIsOpen).toBeFalsy();
    });
  });
  describe('openCancelConfirmationModal method', () => {
    const wrapper = shallow(
      <OrderRow {...props} />
    );
    const action = wrapper.instance();
    it('should set cancelModalIsOpen state to true', () => {
      action.openCancelConfirmationModal();
      const { cancelModalIsOpen } = action.state;

      expect(cancelModalIsOpen).toBeTruthy();
    });
  });
  describe('closeUpdateModal method', () => {
    const wrapper = shallow(
      <OrderRow {...props} />
    );
    const action = wrapper.instance();
    it('should set updateModalIsOpen state to false', () => {
      action.closeUpdateModal();
      const { updateModalIsOpe } = action.state;

      expect(updateModalIsOpe).toBeFalsy();
    });
  });

  describe('openUpdateModal method', () => {
    const wrapper = shallow(
      <OrderRow {...props} />
    );
    const action = wrapper.instance();
    it('should set updateModalIsOpen state to true', () => {
      action.openUpdateModal();
      const { updateModalIsOpen } = action.state;

      expect(updateModalIsOpen).toBeTruthy();
    });
  });
});