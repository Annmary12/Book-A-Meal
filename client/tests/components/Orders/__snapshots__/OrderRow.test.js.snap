// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderRow component should match snapshot if status is anything else 1`] = `
<Row
  header={false}
  key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
>
  <Cell
    title="date"
  >
    9th Sep, 2018
  </Cell>
  <Cell
    title="Meal"
  >
    Beef with rice
  </Cell>
  <Cell
    title="Quantity"
  >
    1
  </Cell>
  <Cell
    title="Price"
  >
    ₦
    500.00
  </Cell>
  <Cell
    title="Delivery address"
  >
    <p
      className="deliveryAddress"
    >
      24 irepodun street, bodija, ibadan
    </p>
  </Cell>
  <Cell
    title="Status"
  >
    <span
      style={
        Object {
          "color": "yellow",
        }
      }
    >
      anything else
    </span>
  </Cell>
  <Cell
    title=""
  />
  <ConfirmCancelOrderModal
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
  <UpdateOrderModal
    handleClose={[Function]}
    isOpen={false}
    order={
      Object {
        "createdAt": "2018-09-10T09:05:53.650Z",
        "deliveryAddress": "24 irepodun street, bodija, ibadan",
        "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
        "meal": Object {
          "caterer": Object {
            "email": "mealbooker@gmail.com",
            "fullName": "Otse Obande",
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "role": "caterer",
            "username": "otseobande",
          },
          "description": "Fried rice and beef",
          "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
          "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
          "price": 500,
          "title": "Beef with rice",
        },
        "phoneNumber": "08131928452",
        "price": 500,
        "quantity": 1,
        "status": "anything else",
        "updatedAt": "2018-09-10T09:06:06.292Z",
      }
    }
  />
</Row>
`;

exports[`OrderRow component should match snapshot if status is cancelled 1`] = `
<Row
  header={false}
  key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
>
  <Cell
    title="date"
  >
    9th Sep, 2018
  </Cell>
  <Cell
    title="Meal"
  >
    Beef with rice
  </Cell>
  <Cell
    title="Quantity"
  >
    1
  </Cell>
  <Cell
    title="Price"
  >
    ₦
    500.00
  </Cell>
  <Cell
    title="Delivery address"
  >
    <p
      className="deliveryAddress"
    >
      24 irepodun street, bodija, ibadan
    </p>
  </Cell>
  <Cell
    title="Status"
  >
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      cancelled
    </span>
  </Cell>
  <Cell
    title=""
  />
  <ConfirmCancelOrderModal
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
  <UpdateOrderModal
    handleClose={[Function]}
    isOpen={false}
    order={
      Object {
        "createdAt": "2018-09-10T09:05:53.650Z",
        "deliveryAddress": "24 irepodun street, bodija, ibadan",
        "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
        "meal": Object {
          "caterer": Object {
            "email": "mealbooker@gmail.com",
            "fullName": "Otse Obande",
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "role": "caterer",
            "username": "otseobande",
          },
          "description": "Fried rice and beef",
          "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
          "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
          "price": 500,
          "title": "Beef with rice",
        },
        "phoneNumber": "08131928452",
        "price": 500,
        "quantity": 1,
        "status": "cancelled",
        "updatedAt": "2018-09-10T09:06:06.292Z",
      }
    }
  />
</Row>
`;

exports[`OrderRow component should match snapshot if status is delivered 1`] = `
<Row
  header={false}
  key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
>
  <Cell
    title="date"
  >
    9th Sep, 2018
  </Cell>
  <Cell
    title="Meal"
  >
    Beef with rice
  </Cell>
  <Cell
    title="Quantity"
  >
    1
  </Cell>
  <Cell
    title="Price"
  >
    ₦
    500.00
  </Cell>
  <Cell
    title="Delivery address"
  >
    <p
      className="deliveryAddress"
    >
      24 irepodun street, bodija, ibadan
    </p>
  </Cell>
  <Cell
    title="Status"
  >
    <span
      style={
        Object {
          "color": "green",
        }
      }
    >
      delivered
    </span>
  </Cell>
  <Cell
    title=""
  />
  <ConfirmCancelOrderModal
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
  <UpdateOrderModal
    handleClose={[Function]}
    isOpen={false}
    order={
      Object {
        "createdAt": "2018-09-10T09:05:53.650Z",
        "deliveryAddress": "24 irepodun street, bodija, ibadan",
        "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
        "meal": Object {
          "caterer": Object {
            "email": "mealbooker@gmail.com",
            "fullName": "Otse Obande",
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "role": "caterer",
            "username": "otseobande",
          },
          "description": "Fried rice and beef",
          "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
          "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
          "price": 500,
          "title": "Beef with rice",
        },
        "phoneNumber": "08131928452",
        "price": 500,
        "quantity": 1,
        "status": "delivered",
        "updatedAt": "2018-09-10T09:06:06.292Z",
      }
    }
  />
</Row>
`;

exports[`OrderRow component should match snapshot if status is not pending 1`] = `
<Row
  header={false}
  key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
>
  <Cell
    title="date"
  >
    9th Sep, 2018
  </Cell>
  <Cell
    title="Meal"
  >
    Beef with rice
  </Cell>
  <Cell
    title="Quantity"
  >
    1
  </Cell>
  <Cell
    title="Price"
  >
    ₦
    500.00
  </Cell>
  <Cell
    title="Delivery address"
  >
    <p
      className="deliveryAddress"
    >
      24 irepodun street, bodija, ibadan
    </p>
  </Cell>
  <Cell
    title="Status"
  >
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      cancelled
    </span>
  </Cell>
  <Cell
    title=""
  />
  <ConfirmCancelOrderModal
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
  <UpdateOrderModal
    handleClose={[Function]}
    isOpen={false}
    order={
      Object {
        "createdAt": "2018-09-10T09:05:53.650Z",
        "deliveryAddress": "24 irepodun street, bodija, ibadan",
        "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
        "meal": Object {
          "caterer": Object {
            "email": "mealbooker@gmail.com",
            "fullName": "Otse Obande",
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "role": "caterer",
            "username": "otseobande",
          },
          "description": "Fried rice and beef",
          "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
          "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
          "price": 500,
          "title": "Beef with rice",
        },
        "phoneNumber": "08131928452",
        "price": 500,
        "quantity": 1,
        "status": "cancelled",
        "updatedAt": "2018-09-10T09:06:06.292Z",
      }
    }
  />
</Row>
`;

exports[`OrderRow component should match snapshot if status is pending 1`] = `
<Row
  header={false}
  key="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
>
  <Cell
    title="date"
  >
    9th Sep, 2018
  </Cell>
  <Cell
    title="Meal"
  >
    Beef with rice
  </Cell>
  <Cell
    title="Quantity"
  >
    1
  </Cell>
  <Cell
    title="Price"
  >
    ₦
    500.00
  </Cell>
  <Cell
    title="Delivery address"
  >
    <p
      className="deliveryAddress"
    >
      24 irepodun street, bodija, ibadan
    </p>
  </Cell>
  <Cell
    title="Status"
  >
    <span
      style={
        Object {
          "color": "#f16800",
        }
      }
    >
      pending
    </span>
  </Cell>
  <Cell
    title=""
  />
  <ConfirmCancelOrderModal
    handleClose={[Function]}
    isOpen={false}
    orderId="ecd2344b-9c49-4c78-8623-f7f3334e2f4e"
  />
  <UpdateOrderModal
    handleClose={[Function]}
    isOpen={false}
    order={
      Object {
        "createdAt": "2018-09-10T09:05:53.650Z",
        "deliveryAddress": "24 irepodun street, bodija, ibadan",
        "id": "ecd2344b-9c49-4c78-8623-f7f3334e2f4e",
        "meal": Object {
          "caterer": Object {
            "email": "mealbooker@gmail.com",
            "fullName": "Otse Obande",
            "id": "e20ac257-86cc-4a6f-a619-0249a201c475",
            "role": "caterer",
            "username": "otseobande",
          },
          "description": "Fried rice and beef",
          "id": "dd72bc9b-beb0-49c6-be7b-20106e8aa3d0",
          "img": "https://qph.ec.quoracdn.net/main-qimg-114f40fc7ad4f623ac45ebb33e2c1f8d",
          "price": 500,
          "title": "Beef with rice",
        },
        "phoneNumber": "08131928452",
        "price": 500,
        "quantity": 1,
        "status": "pending",
        "updatedAt": "2018-09-10T09:06:06.292Z",
      }
    }
  />
</Row>
`;
